packages/core/src/rpc/index.ts

```diff
@@ -45,73 +45,295 @@ export type Rpc = {
-        retryCount: 0,
-        timeout: 5_000,
-      }).request;
+      request = [
+        webSocket(chain.rpc)({
+          chain: chain.viemChain,
+          retryCount: 0,
+          timeout: 5_000,
+        }).request,
+      ];
     } else {
       throw new Error(`Unsupported RPC URL protocol: ${protocol}`);
     }
   } else if (Array.isArray(chain.rpc)) {
-    request = loadBalance(
-      chain.rpc.map((rpc) => {
-        const protocol = new url.URL(rpc).protocol;
-        if (protocol === "https:" || protocol === "http:") {
-          return http(rpc);
-        } else if (protocol === "wss:" || protocol === "ws:") {
-          return webSocket(rpc);
-        } else {
-          throw new Error(`Unsupported RPC URL protocol: ${protocol}`);
-        }
-      }),
-    )({
-      chain: chain.viemChain,
-      retryCount: 0,
-      timeout: 5_000,
-    }).request;
+    request = chain.rpc.map((rpc) => {
+      const protocol = new url.URL(rpc).protocol;
+      if (protocol === "https:" || protocol === "http:") {
+        return http(rpc)({
+          chain: chain.viemChain,
+          retryCount: 0,
+          timeout: 5_000,
+        }).request;
+      } else if (protocol === "wss:" || protocol === "ws:") {
+        return webSocket(rpc)({
+          chain: chain.viemChain,
+          retryCount: 0,
+          timeout: 5_000,
+        }).request;
+      } else {
+        throw new Error(`Unsupported RPC URL protocol: ${protocol}`);
+      }
+    });
   } else {
-    request = chain.rpc({
-      chain: chain.viemChain,
-      retryCount: 0,
-      timeout: 5_000,
-    }).request;
+    request = [
+      chain.rpc({
+        chain: chain.viemChain,
+        retryCount: 0,
+        timeout: 5_000,
+      }).request,
+    ];
   }
 
+  const buckets = request.map(
+    (request, index) =>
+      ({
+        index,
+        reactivationDelay: INITIAL_REACTIVATION_DELAY,
+
+        activeConnections: 0,
+        isActive: true,
+        isWarmingUp: false,
+
+        latencyMetadata: {
+          latencies: [],
+          successfulLatencies: 0,
+          latencySum: 0,
+        },
+        expectedLatency: 200,
+
+        requestTimestamps: [],
+        consecutiveSuccessfulRequests: 0,
+        rpsLimit: INITIAL_MAX_RPS,
+
+        request,
+      }) satisfies Bucket,
+  );
+
+  const scheduleBucketActivation = (bucket: Bucket) => {
+    setTimeout(() => {
+      bucket.isActive = true;
+      bucket.isWarmingUp = true;
+    }, bucket.reactivationDelay);
+  };
+
+  const getBucket = async (): Promise<Bucket> => {
+    const availableBuckets = buckets.filter((b) => isAvailable(b));
+
+    if (availableBuckets.length === 0) {
+      const rpsSafePromise = new Promise<void>((resolve) => {
+        const checkRPSSafe = () => {
+          const availableBucket = buckets.find((b) => isAvailable(b));
+          if (availableBucket) {
+            resolve();
+          } else {
+            setTimeout(checkRPSSafe, 5);
```

Comment:
I wonder if there's a way to do this that doesn't create so many timeouts - IIRC they're somewhat CPU intensive, but I might be wrong.